pkg/clients/redshift/redshift.go

```diff
@@ -0,0 +1,504 @@
+	in.PreferredMaintenanceWindow = awsclients.LateInitializeStringPtr(in.PreferredMaintenanceWindow, cl.PreferredMaintenanceWindow)
+	in.PubliclyAccessible = awsclients.LateInitializeBoolPtr(in.PubliclyAccessible, cl.PubliclyAccessible)
+	in.SnapshotScheduleIdentifier = awsclients.LateInitializeStringPtr(in.SnapshotScheduleIdentifier, cl.SnapshotScheduleIdentifier)
+
+	// If ClusterType is not provided by the user then set it to it's default value.
+	// As redshift.Cluster type doesn't hold this info.
+	if in.ClusterType == nil {
+		if aws.Int64Value(cl.NumberOfNodes) > 1 {
+			in.ClusterType = aws.String("multi-node")
+		}
+		if aws.Int64Value(cl.NumberOfNodes) == 1 {
+			in.ClusterType = aws.String("single-node")
+		}
+	}
+	if cl.Endpoint != nil {
+		in.Port = awsclients.LateInitializeInt64Ptr(in.Port, cl.Endpoint.Port)
+	}
+	if cl.HsmStatus != nil {
+		in.HSMClientCertificateIdentifier = awsclients.LateInitializeStringPtr(in.HSMClientCertificateIdentifier, cl.HsmStatus.HsmClientCertificateIdentifier)
+		in.HSMConfigurationIdentifier = awsclients.LateInitializeStringPtr(in.HSMConfigurationIdentifier, cl.HsmStatus.HsmConfigurationIdentifier)
+	}
+	if cl.ElasticIpStatus != nil {
+		in.ElasticIP = awsclients.LateInitializeStringPtr(in.ElasticIP, cl.ElasticIpStatus.ElasticIp)
+	}
+
+	if len(cl.ClusterSecurityGroups) != 0 {
+		s := make([]string, len(cl.ClusterSecurityGroups))
+		for i, v := range cl.ClusterSecurityGroups {
+			s[i] = aws.StringValue(v.ClusterSecurityGroupName)
+		}
+		in.ClusterSecurityGroups = s
+	}
+	if len(cl.IamRoles) != 0 {
+		s := make([]string, len(cl.IamRoles))
+		for i, v := range cl.IamRoles {
+			s[i] = aws.StringValue(v.IamRoleArn)
+		}
+		in.IAMRoles = s
+	}
+	if len(cl.Tags) != 0 {
+		s := make([]v1alpha1.Tag, len(cl.Tags))
+		for i, v := range cl.Tags {
+			s[i] = v1alpha1.Tag{Key: aws.StringValue(v.Key), Value: aws.StringValue(v.Value)}
+		}
+		in.Tags = s
+	}
+	if len(cl.VpcSecurityGroups) != 0 {
+		s := make([]string, len(cl.VpcSecurityGroups))
+		for i, v := range cl.VpcSecurityGroups {
+			s[i] = aws.StringValue(v.VpcSecurityGroupId)
+		}
+		in.VPCSecurityGroupIDs = s
+	}
+}
+
+// IsUpToDate checks whether there is a change in any of the modifiable fields.
+func IsUpToDate(p v1alpha1.RedshiftParameters, cl redshift.Cluster) (bool, error) { // nolint:gocyclo
+	// We need to check it explicitly as redshift.Cluster can have multiple ClusterParameterGroups
+	found := isClusterParameterGroupNameUpdated(p.ClusterParameterGroupName, cl.ClusterParameterGroups)
+
+	// Check if it is a cluster rename request
+	if p.NewClusterIdentifier != nil && (aws.StringValue(p.NewClusterIdentifier) != aws.StringValue(cl.ClusterIdentifier)) {
+		return false, nil
+	}
+
+	// Since redshift.Cluster doesn't have a ClusterType field therefore determine its value basecd upon number of nodes.
+	if aws.Int64Value(cl.NumberOfNodes) > 1 && aws.StringValue(p.ClusterType) != "multi-node" {
```

Comment:
Do we need to check these explicitly if they are also being late initialized correctly in the `Patch` operation? https://github.com/crossplane/provider-aws/pull/253/files#diff-b96ea1b7a25a962d50f42ee655cc5283R78

Reply:
The idea is to identify the change when the user modifies `ClusterType`  to a valid value. As in `LateInitialize`, we just set it to a sane value, not actual observed value.